### DISCUSSION TITLE
There is a way to change github actions runner location?

### DISCUSSION THREAD

**Discussion by gabrielAOR** [community/community#137547] [Actions] [answered]
I've created a script that retrieves my Binance balance and sends the data to a Google Sheet. I want to use GitHub Actions to run this script every day. However, when the script runs, the Binance library returns an error saying the service is unavailable from the given GitHub Actions server location.

**Comment by behdad088** [comment, marked as answer]
Hi @gabrielAOR 

The error you're encountering is likely due to Binance blocking requests from certain IP addresses, including the IP ranges used by GitHub Actions. Many exchanges and services block traffic from known cloud providers, including GitHub, to prevent abuse or limit access from specific geographic regions.

To resolve this, here are some potential solutions you can explore:

### 1. **Use a VPN or Proxy Service**
   One solution is to route your requests through a VPN or proxy service that Binance allows. You can configure your GitHub Actions workflow to use a VPN or proxy to bypass the restriction.

   **Steps:**
   - Choose a VPN or proxy service that supports API access and whitelist requests from its IP ranges.
   - Update your script to make requests through the proxy or VPN.
   - Set up your GitHub Actions workflow to include the necessary environment variables or proxy configuration for the VPN.

   Example of how to add a proxy to your script:
   ```python
   import requests

   proxy = "http://yourproxy.com:8080"
   response = requests.get('https://api.binance.com/api/v3/account', proxies={"http": proxy, "https": proxy})
   ```

### 2. **Self-Hosted GitHub Runner**
   A better long-term solution might be to run a self-hosted GitHub Actions runner on a server that you control and that Binance doesn't block. This will allow you to run the script without relying on GitHub's IP ranges.

   **Steps:**
   - Set up a self-hosted runner on a VPS or a machine in a location that Binance allows.
   - Register the runner with GitHub, and modify your GitHub Actions workflow to use that self-hosted runner.

   [GitHub documentation](https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners) provides instructions for setting up a self-hosted runner.

### 3. **Use a Serverless Function (e.g., AWS Lambda, Google Cloud Functions)**
   Another approach is to deploy your script as a serverless function using services like AWS Lambda, Google Cloud Functions, or Azure Functions. You can schedule these functions to run daily using their respective schedulers (e.g., AWS CloudWatch, Google Cloud Scheduler).

   **Steps for AWS Lambda Example:**
   - Move your Binance balance retrieval script to an AWS Lambda function.
   - Use AWS CloudWatch Events to trigger the function daily.
   - Ensure that the IP range used by AWS Lambda is allowed by Binance.

   This method moves the execution environment out of GitHub Actions entirely, allowing you to schedule it independently.

### 4. **Run the Script Locally with a Cron Job**
   If you're okay with running the script from a local machine, you can set up a cron job on your computer or server to execute the script daily.

   **Steps:**
   - Set up a cron job on a local machine or server to run the script every day.
   - Make sure the machine is located in a region allowed by Binance.
   - You can still push the results to GitHub or Google Sheets.

   Example cron job entry:
   ```
   0 0 * * * /usr/bin/python3 /path/to/your/script.py
   ```

### 5. **Use Another CI/CD Service**
   If you're committed to using a CI/CD service but GitHub Actions is being blocked by Binance, you can explore other CI/CD platforms like CircleCI, Travis CI, or GitLab CI. Each of these services has different IP ranges, and it's possible Binance doesn't block them.

   **Steps:**
   - Move your script to another CI/CD service.
   - Schedule the script to run daily using that service's scheduling tools.

   If the other services' IP ranges are also blocked, you can combine this approach with using a proxy or VPN service as outlined in the first solution.

If you prefer a hands-off solution, the self-hosted runner or serverless functions approach might be the best way to go.
